sprs/src/sparse/csmat.rs

```diff
@@ -549,18 +549,38 @@ impl<N, I: SpIndex, Iptr: SpIndex> CsMatI<N, I, Iptr> {
     }
 
     /// Append an outer dim to an existing matrix, compressing it in the process
-    pub fn append_outer(mut self, data: &[N]) -> Self
+    pub fn append_outer(self, data: &[N]) -> Self
     where
         N: Clone + Num,
+    {
+        self.append_outer_iter(data.iter().cloned().enumerate())
+    }
+
+    /// Append an outer dim to an existing matrix, provided by an iterator
+    pub fn append_outer_iter<Iter>(mut self, iter: Iter) -> Self
+    where
+        N: Num,
+        Iter: Iterator<Item = (usize, N)>,
     {
         let mut nnz = self.nnz();
-        for (inner_ind, val) in data.iter().enumerate() {
-            if *val != N::zero() {
+        let mut prev_inner_ind = None;
+        for (inner_ind, val) in iter {
+            if val != N::zero() {
+                if let Some(p_ind) = prev_inner_ind {
+                    assert!(p_ind < inner_ind, "inner index order");
```

Comment:
```suggestion
                    assert!(p_ind < inner_ind, "Indices in provided data are out of order. To fix this, sort by index in ascending order.");
```

Reply:
I'm personally not a big fan of elaborate error-messages. Perhaps provide some data to help out instead? What about something like
```rust
      assert!(p_ind < inner_ind, "Indices in provided data are out of order. Got {} then {}", p_ind, inner_ind);
```